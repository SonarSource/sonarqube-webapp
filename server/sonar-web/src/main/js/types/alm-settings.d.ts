/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
declare namespace T {
  export interface AlmSettingsBinding {
    key: string;
  }

  export interface AlmSettingsInstance {
    alm: string;
    key: string;
    url?: string;
  }

  export interface AlmSettingsBindingDefinitions {
    azure: AzureBindingDefinition[];
    github: GithubBindingDefinition[];
  }

  export interface GithubBindingDefinition extends AlmSettingsBinding {
    url: string;
    appId: string;
    privateKey: string;
  }

  export interface AzureBindingDefinition extends AlmSettingsBinding {
    personalAccessToken: string;
  }

  export interface ProjectAlmBinding {
    key: string;
    repository?: string;
  }

  export interface AzureProjectAlmBinding {
    almSetting: string;
    project: string;
  }

  export interface GithubProjectAlmBinding {
    almSetting: string;
    project: string;
    repository: string;
  }
}
