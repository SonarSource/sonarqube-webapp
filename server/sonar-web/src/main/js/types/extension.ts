/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { IntlShape } from 'react-intl';
import { Store as ReduxStore } from 'redux';
import { Location, Router } from '../components/hoc/withRouter';
import { Store } from '../store/rootReducer';
import { L10nBundle } from './l10n';
import { AppState, CurrentUser, Dict } from './types';

export enum AdminPageExtension {
  GovernanceConsole = 'governance/views_console'
}

export interface ExtensionRegistryEntry {
  start: ExtensionStartMethod;
  providesCSSFile: boolean;
}

export interface ExtensionStartMethod {
  (params: ExtensionStartMethodParameter | string): ExtensionStartMethodReturnType;
}

export interface ExtensionStartMethodParameter {
  appState: AppState;
  store: ReduxStore<Store, any>;
  el: HTMLElement | undefined | null;
  currentUser: CurrentUser;
  intl: IntlShape;
  location: Location;
  router: Router;
  theme: {
    colors: Dict<string>;
    sizes: Dict<string>;
    rawSizes: Dict<number>;
    fonts: Dict<string>;
    zIndexes: Dict<string>;
    others: Dict<string>;
  };
  baseUrl: string;
  l10nBundle: L10nBundle;
}

export type ExtensionStartMethodReturnType = React.ReactNode | Function | void | undefined | null;
