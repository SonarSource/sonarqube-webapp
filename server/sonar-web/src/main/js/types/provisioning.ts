/*
 * SonarQube
 * Copyright (C) 2009-2023 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { TaskStatuses } from './tasks';

export type GithubStatusDisabled = {
  enabled: false;
  nextSync?: never;
  lastSync?: never;
};
export interface GithubStatusEnabled extends AlmSyncStatus {
  enabled: true;
}

export interface AlmSyncStatus {
  nextSync?: { status: TaskStatuses.Pending | TaskStatuses.InProgress };
  lastSync?: {
    executionTimeMs: number;
    finishedAt: number;
    startedAt: number;
    warningMessage?: string;
  } & (
    | {
        status: TaskStatuses.Success;
        summary?: string;
        errorMessage?: never;
      }
    | {
        status: TaskStatuses.Canceled | TaskStatuses.Failed;
        summary?: never;
        errorMessage?: string;
      }
  );
}

export type GithubStatus = GithubStatusDisabled | GithubStatusEnabled;

export enum GitHubProvisioningStatus {
  Success = 'SUCCESS',
  Failed = 'FAILED',
}

type GitHubProvisioning =
  | {
      status: GitHubProvisioningStatus.Success;
      errorMessage?: never;
    }
  | {
      status: GitHubProvisioningStatus.Failed;
      errorMessage: string;
    };

export interface GitHubConfigurationStatus {
  application: {
    jit: GitHubProvisioning;
    autoProvisioning: GitHubProvisioning;
  };
  installations: {
    organization: string;
    jit: GitHubProvisioning;
    autoProvisioning: GitHubProvisioning;
  }[];
}

export interface GitHubMapping {
  readonly id: string;
  readonly githubRole: string;
  readonly isBaseRole?: boolean;
  permissions: {
    user: boolean;
    codeViewer: boolean;
    issueAdmin: boolean;
    securityHotspotAdmin: boolean;
    admin: boolean;
    scan: boolean;
  };
}

export interface GitLabConfigurationCreateBody {
  applicationId: string;
  url: string;
  clientSecret: string;
  synchronizeUserGroups: boolean;
}

export type GitLabConfigurationUpdateBody = {
  applicationId?: string;
  url?: string;
  clientSecret?: string;
  synchronizeUserGroups?: boolean;
  enabled?: boolean;
  type?: ProvisioningType;
  provisioningToken?: string;
  groups?: string[];
  allowUsersToSignUp?: boolean;
};

export type GitlabConfiguration = {
  id: string;
  enabled: boolean;
  synchronizeUserGroups: boolean;
  url: string;
  type: ProvisioningType;
  groups: string[];
  allowUsersToSignUp: boolean;
};

export enum ProvisioningType {
  jit = 'JIT',
  auto = 'Auto',
}
