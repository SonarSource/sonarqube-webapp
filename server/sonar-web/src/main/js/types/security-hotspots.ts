/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export enum RiskExposure {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH'
}

export enum HotspotStatus {
  TO_REVIEW = 'TO_REVIEW',
  REVIEWED = 'REVIEWED'
}

export enum HotspotResolution {
  FIXED = 'FIXED',
  SAFE = 'SAFE'
}

export enum HotspotStatusOptions {
  FIXED = 'FIXED',
  SAFE = 'SAFE',
  ADDITIONAL_REVIEW = 'ADDITIONAL_REVIEW'
}

export interface RawHotspot {
  assignee?: string;
  author?: string;
  component: string;
  creationDate: string;
  key: string;
  line?: number;
  message: string;
  project: string;
  resolution?: string;
  rule: string;
  securityCategory: string;
  status: string;
  subProject?: string;
  updateDate: string;
  vulnerabilityProbability: RiskExposure;
}

export interface DetailedHotspot {
  assignee?: Pick<T.UserBase, 'active' | 'login' | 'name'>;
  author?: Pick<T.UserBase, 'login'>;
  component: T.Component;
  creationDate: string;
  key: string;
  line?: number;
  message: string;
  project: T.Component;
  resolution?: string;
  rule: DetailedHotspotRule;
  status: string;
  textRange: T.TextRange;
  updateDate: string;
}

export interface HotspotUpdateFields {
  status: HotspotStatus;
  resolution?: HotspotResolution;
}

export interface HotspotUpdate extends HotspotUpdateFields {
  key: string;
}

export interface DetailedHotspotRule {
  fixRecommendations?: string;
  key: string;
  name: string;
  riskDescription?: string;
  securityCategory: string;
  vulnerabilityDescription?: string;
  vulnerabilityProbability: RiskExposure;
}

export interface HotspotSearchResponse {
  components?: { key: string; qualifier: string; name: string }[];
  hotspots: RawHotspot[];
  paging: T.Paging;
}

export interface HotspotSetStatusRequest {
  hotspot: string;
  status: HotspotStatus;
  resolution?: HotspotResolution;
}
