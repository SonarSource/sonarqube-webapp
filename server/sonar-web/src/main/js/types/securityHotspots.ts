/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export enum RiskExposure {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH'
}

export interface RawHotspot {
  assignee?: string;
  author?: string;
  component: string;
  creationDate: string;
  key: string;
  line?: number;
  message: string;
  project: string;
  resolution: string;
  rule: string;
  securityCategory: string;
  updateDate: string;
  vulnerabilityProbability: RiskExposure;
  status: string;
  subProject?: string;
}

export interface HotspotSearchResponse {
  components?: { key: string; qualifier: string; name: string }[];
  hotspots: RawHotspot[];
  paging: T.Paging;
}
