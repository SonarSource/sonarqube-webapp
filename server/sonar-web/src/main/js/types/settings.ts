/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Dict } from './types';

export const enum SettingsKey {
  DaysBeforeDeletingInactiveBranchesAndPRs = 'sonar.dbcleaner.daysBeforeDeletingInactiveBranchesAndPRs',
  DefaultProjectVisibility = 'projects.default.visibility',
  ServerBaseUrl = 'sonar.core.serverBaseURL',
  PluginRiskConsent = 'sonar.plugins.risk.consent'
}

export type SettingDefinitionAndValue = {
  definition: ExtendedSettingDefinition;
  settingValue?: SettingValue;
};

export type Setting = SettingValue & { definition: SettingDefinition; hasValue: boolean };
export type SettingWithCategory = Setting & { definition: ExtendedSettingDefinition };

export enum SettingType {
  STRING = 'STRING',
  TEXT = 'TEXT',
  JSON = 'JSON',
  PASSWORD = 'PASSWORD',
  BOOLEAN = 'BOOLEAN',
  FLOAT = 'FLOAT',
  INTEGER = 'INTEGER',
  LICENSE = 'LICENSE',
  LONG = 'LONG',
  SINGLE_SELECT_LIST = 'SINGLE_SELECT_LIST',
  PROPERTY_SET = 'PROPERTY_SET'
}
export interface SettingDefinition {
  description?: string;
  key: string;
  multiValues?: boolean;
  name?: string;
  options: string[];
  type?: SettingType;
}

export interface SettingFieldDefinition extends SettingDefinition {
  description: string;
  name: string;
}

export interface ExtendedSettingDefinition extends SettingDefinition {
  category: string;
  defaultValue?: string;
  deprecatedKey?: string;
  fields: SettingFieldDefinition[];
  multiValues?: boolean;
  subCategory: string;
}

export interface SettingValueResponse {
  settings: SettingValue[];
  setSecuredSettings: string[];
}

export interface SettingValue {
  fieldValues?: Array<Dict<string>>;
  inherited?: boolean;
  key: string;
  parentFieldValues?: Array<Dict<string>>;
  parentValue?: string;
  parentValues?: string[];
  value?: string;
  values?: string[];
}
