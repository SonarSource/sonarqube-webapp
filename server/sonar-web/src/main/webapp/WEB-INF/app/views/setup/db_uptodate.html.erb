<meta http-equiv='refresh' content='5;url=<%= home_path -%>'>
<h1 class="maintenance-title text-success">Database is up-to-date</h1>
<p class="maintenance-text">
  You will be redirected shortly to SonarQube.
  <br>
  <small class="text-muted">
    (if this does not happen, you can <a href="<%= home_path -%>">click here to be redirected</a>)
  </small>
</p>
