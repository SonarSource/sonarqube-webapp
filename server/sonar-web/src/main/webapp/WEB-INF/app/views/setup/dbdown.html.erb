<h1 class="maintenance-title text-danger">Fail to connect to database</h1>
<p class="maintenance-text">Database connection cannot be established. Please check database status and JDBC settings.</p>
<div class="maintenance-spinner">
  <%= button_to "Try again", { :action => "index" }, :method => :get %>
</div>
