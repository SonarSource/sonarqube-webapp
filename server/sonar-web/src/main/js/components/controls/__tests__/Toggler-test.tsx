/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { act, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup';
import * as React from 'react';
import { byRole } from 'testing-library-selector';
import Toggler from '../Toggler';

beforeAll(() => {
  jest.useFakeTimers();
});

afterAll(() => {
  jest.useRealTimers();
});
const ui = {
  toggleButton: byRole('button', { name: 'toggle' }),
  outButton: byRole('button', { name: 'out' }),
  overlayButton: byRole('button', { name: 'overlay' }),
};

async function openToggler(user: UserEvent) {
  await user.click(ui.toggleButton.get());
  act(() => {
    jest.runAllTimers();
  });
  expect(ui.overlayButton.get()).toBeInTheDocument();
}

function focusOut() {
  act(() => {
    ui.overlayButton.get().focus();
    ui.outButton.get().focus();
  });
}

it('should handle escape correclty', async () => {
  const user = userEvent.setup({ delay: null });
  const rerender = renderToggler({
    closeOnEscape: true,
    closeOnClick: false,
    closeOnClickOutside: false,
    closeOnFocusOut: false,
  });

  await openToggler(user);

  await user.keyboard('{Escape}');
  expect(ui.overlayButton.query()).not.toBeInTheDocument();

  rerender({ closeOnEscape: false });
  await openToggler(user);

  await user.keyboard('{Escape}');
  expect(ui.overlayButton.get()).toBeInTheDocument();
});

it('should handle focus correctly', async () => {
  const user = userEvent.setup({ delay: null });
  const rerender = renderToggler({
    closeOnEscape: false,
    closeOnClick: false,
    closeOnClickOutside: false,
    closeOnFocusOut: true,
  });

  await openToggler(user);

  focusOut();
  expect(ui.overlayButton.query()).not.toBeInTheDocument();

  rerender({ closeOnFocusOut: false });
  await openToggler(user);

  focusOut();
  expect(ui.overlayButton.get()).toBeInTheDocument();
});

it('should handle click correctly', async () => {
  const user = userEvent.setup({ delay: null });
  const rerender = renderToggler({
    closeOnEscape: false,
    closeOnClick: true,
    closeOnClickOutside: false,
    closeOnFocusOut: false,
  });

  await openToggler(user);

  await user.click(ui.outButton.get());
  expect(ui.overlayButton.query()).not.toBeInTheDocument();

  await openToggler(user);

  await user.click(ui.overlayButton.get());
  expect(ui.overlayButton.query()).not.toBeInTheDocument();

  rerender({ closeOnClick: false });
  await openToggler(user);

  await user.click(ui.outButton.get());
  expect(ui.overlayButton.get()).toBeInTheDocument();
});

it('should handle click outside correctly', async () => {
  const user = userEvent.setup({ delay: null });
  const rerender = renderToggler({
    closeOnEscape: false,
    closeOnClick: false,
    closeOnClickOutside: true,
    closeOnFocusOut: false,
  });

  await openToggler(user);

  await user.click(ui.overlayButton.get());
  expect(await ui.overlayButton.find()).toBeInTheDocument();

  await user.click(ui.outButton.get());
  expect(ui.overlayButton.query()).not.toBeInTheDocument();

  rerender({ closeOnClickOutside: false });
  await openToggler(user);

  await user.click(ui.outButton.get());
  expect(ui.overlayButton.get()).toBeInTheDocument();
});

it('should open/close correctly when default props is applied', async () => {
  const user = userEvent.setup({ delay: null });
  renderToggler();

  await openToggler(user);

  // Should not close when on overlay
  await user.click(ui.overlayButton.get());
  expect(await ui.overlayButton.find()).toBeInTheDocument();

  // Focus out should close
  act(() => {
    ui.overlayButton.get().focus();
    ui.outButton.get().focus();
  });
  expect(ui.overlayButton.query()).not.toBeInTheDocument();

  await openToggler(user);

  // Escape should close
  await user.keyboard('{Escape}');
  expect(ui.overlayButton.query()).not.toBeInTheDocument();

  await openToggler(user);

  // Click should close (focus out is trigger first)
  await user.click(ui.outButton.get());
  expect(ui.overlayButton.query()).not.toBeInTheDocument();
});

function renderToggler(override?: Partial<Toggler['props']>) {
  function App(props: Partial<Toggler['props']>) {
    const [open, setOpen] = React.useState(false);

    return (
      <>
        <Toggler
          onRequestClose={() => setOpen(false)}
          open={open}
          overlay={<button type="button">overlay</button>}
          {...props}
        >
          <button onClick={() => setOpen(true)} type="button">
            toggle
          </button>
        </Toggler>
        <button type="button">out</button>
      </>
    );
  }

  const { rerender } = render(<App {...override} />);
  return function (reoverride: Partial<Toggler['props']>) {
    return rerender(<App {...override} {...reoverride} />);
  };
}
