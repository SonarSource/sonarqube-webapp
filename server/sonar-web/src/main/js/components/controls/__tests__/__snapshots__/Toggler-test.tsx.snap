// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render click wrappers 1`] = `
<React.Fragment>
  <div
    id="toggle"
  />
  <div
    id="overlay"
  />
</React.Fragment>
`;

exports[`should render children and overlay 1`] = `
<React.Fragment>
  <div
    id="toggle"
  />
  <OutsideClickHandler
    onClickOutside={[MockFunction]}
  >
    <div
      id="overlay"
    />
  </OutsideClickHandler>
</React.Fragment>
`;

exports[`should render only children 1`] = `
<React.Fragment>
  <div
    id="toggle"
  />
</React.Fragment>
`;

exports[`should render when closeOnClick=true 1`] = `
<React.Fragment>
  <div
    id="toggle"
  />
  <DocumentClickHandler
    onClick={[MockFunction]}
  >
    <div
      id="overlay"
    />
  </DocumentClickHandler>
</React.Fragment>
`;
